### PR TITLE
[Filebeat] Add support for events triggered by S3 uploads from the AWS console in S3 input

### PR DESCRIPTION
**Pull request by darupedk**
- Add support for the `ObjectCreated:CompleteMultipartUpload` event in S3 input

### CODE REVIEW

**Comment by kaiyan-sheng** [comment]
jenkins, test this please

**Comment by kaiyan-sheng** [comment]
@darupedk Thanks for adding this 👍  We are working on adding elb fileset in Filebeat aws module and ran into this same problem. 

Looks like this PR might need a rebase to make CI happy. Could you do that please?

**Comment by darupedk** [comment]
> @darupedk Thanks for adding this 👍 We are working on adding elb fileset in Filebeat aws module and ran into this same problem.
> 
> Looks like this PR might need a rebase to make CI happy. Could you do that please?

There we go.

In case we want full support for create events, we need to support the complete list of events:

- s3:ObjectCreated:Put
- s3:ObjectCreated:Post
- s3:ObjectCreated:Copy
- s3:ObjectCreated:CompleteMultipartUpload

I haven't inspected the Post and Copy events (JSON) which is why I haven't added them. However, we could do `strings.HasPrefix('s3:ObjectCreated:')` and check if the Post and Copy events are based on the same JSON structure. I wouldn't mind checking up on that.

**Comment by darupedk** [comment]
> > @darupedk Thanks for adding this 👍 We are working on adding elb fileset in Filebeat aws module and ran into this same problem.
> > Looks like this PR might need a rebase to make CI happy. Could you do that please?
> 
> There we go.
> 
> In case we want full support for create events, we need to support the complete list of events:
> 
> * s3:ObjectCreated:Put
> * s3:ObjectCreated:Post
> * s3:ObjectCreated:Copy
> * s3:ObjectCreated:CompleteMultipartUpload
> 
> I haven't inspected the Post and Copy events (JSON) which is why I haven't added them. However, we could do `strings.HasPrefix('s3:ObjectCreated:')` and check if the Post and Copy events are based on the same JSON structure. I wouldn't mind checking up on that.

I have just inspected them and they do follow the same structure. I recommend changing the code to use `strings.HasPrefix('s3:ObjectCreated:')`.

File: x-pack/filebeat/input/s3/input.go [hunk 1]
```diff
@@ -340,7 +340,7 @@ func handleSQSMessage(m sqs.Message) ([]s3Info, error) {
 
 	var s3Infos []s3Info
 	for _, record := range msg.Records {
-		if record.EventSource == "aws:s3" && record.EventName == "ObjectCreated:Put" {
+		if record.EventSource == "aws:s3" && strings.HasPrefix(record.EventName, "ObjectCreated:") {
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/s3/input.go]: How about we use the constant `EventS3ObjectCreated` here from aws-sdk-go-v2 https://github.com/aws/aws-sdk-go-v2/blob/master/service/s3/api_enums.go#L2?

Comment (darupedk) [reply x-pack/filebeat/input/s3/input.go]: > How about we use the constant `EventS3ObjectCreated` here from aws-sdk-go-v2 https://github.com/aws/aws-sdk-go-v2/blob/master/service/s3/api_enums.go#L2?

The enums there is the full event name including the `s3:` prefix. However, the raw event data looks like this:

```
{
   "Records":[ 
      { 
         "eventVersion":"2.1",
         "eventSource":"aws:s3",
         "awsRegion":"eu-central-1",
         "eventTime":"2019-10-16T18:20:33.017Z",
         "eventName":"ObjectCreated:CompleteMultipartUpload",
         "userIdentity":{ 
            "principalId":"AWS:123"
         },
         "requestParameters":{ 
            "sourceIPAddress":"0.0.0.0"
         },
         "responseElements":{ 
            "x-amz-request-id":"ABCD",
            "x-amz-id-2":"abc="
         },
         "s3":{ 
            "s3SchemaVersion":"1.0",
            "configurationId":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
            "bucket":{ 
               "name":"testbucket",
               "ownerIdentity":{ 
                  "principalId":"ABCD"
               },
               "arn":"arn:aws:s3:::testbucket"
            },
            "object":{ 
               "key":"accesslog_20191009-132000-00.gz",
               "size":8192,
               "eTag":"abcd",
               "versionId":"abcd",
               "sequencer":"AADA50XXXX"
            }
         }
      }
   ]
}

Which means that we cannot use the enums directly unless we use a substring match or string manipulation. That's why I recommend using HasPrefix.
```

Comment (kaiyan-sheng) [reply x-pack/filebeat/input/s3/input.go]: Ahh good point, didn't see the `s3:` prefix. Thanks!

**Comment by kaiyan-sheng** [comment]
jenkins, test this please

**Comment by darupedk** [comment]
TestHandleMessage is failing due to the following invalid S3 event:
```
		{
			"sqs message with event source aws:s3 and event name ObjectCreated:Delete",
			sqs.Message{
				Body: awssdk.String("{\"Records\":[{\"eventSource\":\"aws:s3\",\"awsRegion\":\"ap-southeast-1\",\"eventTime\":\"2019-06-21T16:16:54.629Z\",\"eventName\":\"ObjectCreated:Delete\",\"s3\":{\"configurationId\":\"object-created-event\",\"bucket\":{\"name\":\"test-s3-ks-2\",\"arn\":\"arn:aws:s3:::test-s3-ks-2\"},\"object\":{\"key\":\"server-access-logging2019-06-21-16-16-54-E68E4316CEB285AA\"}}}]}"),
			},
			[]s3Info{},
		},
```
We can go ahead and validate the events a bit more in the code but AWS would never send a message like this. We could instead replace the test message with `ObjectRemoved:Delete` which is a valid S3 event though not one that should be handled by the plugin.

**Comment by kaiyan-sheng** [comment]
> TestHandleMessage is failing due to the following invalid S3 event:
> 
> ```
> 		{
> 			"sqs message with event source aws:s3 and event name ObjectCreated:Delete",
> 			sqs.Message{
> 				Body: awssdk.String("{\"Records\":[{\"eventSource\":\"aws:s3\",\"awsRegion\":\"ap-southeast-1\",\"eventTime\":\"2019-06-21T16:16:54.629Z\",\"eventName\":\"ObjectCreated:Delete\",\"s3\":{\"configurationId\":\"object-created-event\",\"bucket\":{\"name\":\"test-s3-ks-2\",\"arn\":\"arn:aws:s3:::test-s3-ks-2\"},\"object\":{\"key\":\"server-access-logging2019-06-21-16-16-54-E68E4316CEB285AA\"}}}]}"),
> 			},
> 			[]s3Info{},
> 		},
> ```
> 
> We can go ahead and validate the events a bit more in the code but AWS would never send a message like this. We could instead replace the test message with `ObjectRemoved:Delete` which is a valid S3 event though not one that should be handled by the plugin.

Changing to `ObjectRemoved:Delete` sounds good. Sorry I made this up just for pure testing purpose originally.

File: x-pack/filebeat/input/s3/input.go (hunk 1 above)
Comment (kaiyan-sheng) [reply x-pack/filebeat/input/s3/input.go]: @jsoriano Since we can't use the enum directly, checking the prefix sounds good to me. WDYT?

Comment (jsoriano) [reply x-pack/filebeat/input/s3/input.go]: Sounds great, thanks!

**Comment by kaiyan-sheng** [comment]
> LGTM, we will need to backport this

Thanks! I will merge it now and backport it.

**Pull request merged**